src/segment_reshape/topology/find_related.py

```diff
@@ -122,49 +134,45 @@ def _as_point_or_line_components(geom: QgsGeometry) -> Iterator[QgsGeometry]:
             yield QgsGeometry(part.clone())
 
 
-def _get_containing_component(
-    geom: QgsGeometry, contains_geom: QgsGeometry
-) -> QgsGeometry:
-    for component in _as_point_or_line_components(geom):
-        if component.contains(contains_geom):
-            return component
+def _get_geom_component_of_vertex(geom: QgsGeometry, vertex_id: int) -> QgsGeometry:
+    """Returns the geometry component that vertex_id belongs to.
 
-    raise ValueError(
-        f"could not find component of {geom} that contains {contains_geom}"
-    )
+    Geometry component is either a single point or linestring from multipoint or
+    multilinestring. In case of polygons geometry component is the edge linear ring
+    that verted_id belongs to.
+
+    Args:
+        geom (QgsGeometry): Input geometry
+        vertex_id (int): Id of the vertex
 
+    Raises:
+        ValueError: Raised if vertex_id is not found from the input geometry.
+
+    Returns:
+        QgsGeometry: The geometry component vertex_id belongs to
+    """
+
+    success, vertex_details = geom.vertexIdFromVertexNr(vertex_id)
+    if not success:
+        raise ValueError(
+            f"could not find vertex details by index {vertex_id}" f" from {geom}"
+        )
 
-def _get_line_component_for_part_and_ring(
-    geom: QgsGeometry, part_index: int, ring_index: int
-) -> QgsGeometry:
     try:
-        part = list(geom.constParts())[part_index]
+        part = list(geom.constParts())[vertex_details.part]
         if not isinstance(part, QgsPolygon):
             return QgsGeometry(part.clone())
-        elif ring_index == 0:
+        elif vertex_details.ring == 0:
             return QgsGeometry(part.exteriorRing().clone())
         else:
-            return QgsGeometry(part.interiorRing(ring_index - 1).clone())
+            return QgsGeometry(part.interiorRing(vertex_details.ring - 1).clone())
     except IndexError:
         raise ValueError(
-            f"could not extract part {part_index} ring {ring_index} of {geom}"
+            f"Could not extract part {vertex_details.part} ring {vertex_details.ring} "
+            f"of {geom}. If happens this is a malfunction in vertexIdFromVertexNr."
         )
```

Comment:
Do we need this exception (or try catch) if this would be actually an issue in vertexIdFromVertexNr? Could these just be handled with qgis_plugin_tools log_if_fails-wrapper in maptool and it would be shown at message bar as unhandled exception?

Reply:
I thought also is this needed at all but didn't want to leave unhandled exception here.

Would the `_handle_mouse_click_event` method to be a good place for `log_if_fails`?

I noticed that currently none of these raised exceptations are handled.

Reply:
Oh, oops. Yes, I think `_handle_mouse_click_event` is a good place for `log_if_fails`